### ISSUE TITLE
gcc-13.2.0 unknown type name '_Float32'

### ISSUE THREAD

**Issue by Nic30** [CastXML/CastXML#251]
From some reason float types do not get defined for gcc-13. With gcc-12 it runs just fine. (and it also works with gcc-13 which faked gcc version, as shown at the end of this post)

`$ "/usr/bin/castxml" -std=c++17 -c -x c++ --castxml-cc-gnu  "(" /usr/bin/c++  -std=c++17  ")" --castxml-gccxml -o /tmp/tmp0.xml /usr/include/stdlib.h
`
```
/usr/include/stdlib.h:141:8: error: unknown type name '_Float32'
extern _Float32 strtof32 (const char *__restrict __nptr,
       ^
...
```

```cpp
// /usr/include/stdlib.h:141
#if __HAVE_FLOAT32 && __GLIBC_USE (IEC_60559_TYPES_EXT)
extern _Float32 strtof32 (const char *__restrict __nptr,
			  char **__restrict __endptr)
```

```cpp
// floatn-common.h

#define __HAVE_FLOAT32 1
...

# if __HAVE_FLOAT32

#  if !__GNUC_PREREQ (7, 0) || (defined __cplusplus && !__GNUC_PREREQ (13, 0))
typedef float _Float32;
#  endif
...
```

```cpp
// features.h

#if defined __GNUC__ && defined __GNUC_MINOR__
# define __GNUC_PREREQ(maj, min) \
	((__GNUC__ << 16) + __GNUC_MINOR__ >= ((maj) << 16) + (min))
#else
# define __GNUC_PREREQ(maj, min) 0
#endif
```

```
$ lsb_release -a
Description:	Ubuntu 23.10
$ /usr/bin/c++ --version
c++ (Ubuntu 13.2.0-4ubuntu3) 13.2.0
$ uname -a
Linux nic30-Precision-5530 6.5.0-14-generic #14-Ubuntu SMP PREEMPT_DYNAMIC Tue Nov 14 14:59:49 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
$ ldd --version
ldd (Ubuntu GLIBC 2.38-1ubuntu6) 2.38
$ castxml -v
Ubuntu clang version 14.0.6
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: 
Found candidate GCC installation: /../lib/gcc/x86_64-linux-gnu/11
Found candidate GCC installation: /../lib/gcc/x86_64-linux-gnu/12
Found candidate GCC installation: /../lib/gcc/x86_64-linux-gnu/13
Selected GCC installation: /../lib/gcc/x86_64-linux-gnu/13
Candidate multilib: .;@m64
Candidate multilib: 32;@m32
Candidate multilib: x32;@mx32
Selected multilib: .;@m64

```

If I fake gcc version everything seems to be working.
```
"/usr/bin/castxml" -std=c++17 -Wno-unknown-attributes -c -x c++ --castxml-cc-gnu  "(" /usr/bin/g++-13  -std=c++17 -D__GNUC__=12 -D__GNUC_MINOR__=0 -Wno-unknown-attributes  ")" --castxml-gccxml -o /tmp/tmp0.xml /usr/include/stdlib.h
```

But there are problems related to gcc version in other important files, e.g.:
```
/usr/include/c++/13/bits/new_allocator.h:163:6: error: call to '__builtin_operator_delete' selects non-usual deallocation function
            _GLIBCXX_OPERATOR_DELETE(_GLIBCXX_SIZED_DEALLOC(__p, __n),
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/13/bits/new_allocator.h:113:35: note: expanded from macro '_GLIBCXX_OPERATOR_DELETE'
define _GLIBCXX_OPERATOR_DELETE __builtin_operator_delete
                                  ^
/usr/include/c++/13/bits/alloc_traits.h:516:13: note: in instantiation of member function 'std::__new_allocator<char>::deallocate' requested here
      { __a.deallocate(__p, __n); }
```

So this can not be used as a hotfix.

If I define all float types manually it seems to be working, but there is another issue with `initializer_list`, which can be solved by `#include <initializer_list>`
```
"/usr/bin/castxml" -std=c++17 -Wno-unknown-attributes -c -x c++ --castxml-cc-gnu  "(" /usr/bin/g++-13  -std=c++17 -D_Float32=float -D_Float64=double -D_Float128=__float128 -D_Float32x=double -D_Float64x="long double" -Wno-unknown-attributes  ")" --castxml-gccxml -o /tmp/tmp0.xml /usr/include/c++/13/bits/hashtable_policy.h
/usr/include/c++/13/bits/hashtable_policy.h:963:14: error: no template named 'initializer_list'
      insert(initializer_list<value_type> __l)
             ^
1 error generated.
```

Fixing previous issue leads to an issue with sized deallocations, which can be fixed with `-fsized-deallocation`.
```
In file included from /usr/include/x86_64-linux-gnu/c++/13/bits/c++allocator.h:33:
/usr/include/c++/13/bits/new_allocator.h:163:6: error: call to '__builtin_operator_delete' selects non-usual deallocation function
            _GLIBCXX_OPERATOR_DELETE(_GLIBCXX_SIZED_DEALLOC(__p, __n),
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/13/bits/new_allocator.h:113:35: note: expanded from macro '_GLIBCXX_OPERATOR_DELETE'
# define _GLIBCXX_OPERATOR_DELETE __builtin_operator_delete
```

So in my case if I want to parse files of ANTL4 cpp runtime library I need to use these options to hotfix all mentioned issues.
```
-std=c++17 -fsized-deallocation -D_Float32=float -D_Float64=double -D_Float128=__float128 -D_Float32x=double -D_Float64x="long double"
```

**Comment by bradking** [comment]
The underlying problem is that with `--castxml-cc-gnu c++` we simulate the real GNU compiler's preprocessor, but the LLVM/Clang on which CastXML is built does not actually have the `_Float32` or similar builtin types.  Similar problems have come up before.  I'll look into possible work arounds.

Also, thanks for the initial analysis.  The condition `!__GNUC_PREREQ (7, 0) || (defined __cplusplus && !__GNUC_PREREQ (13, 0))` looks like we should have been getting this error in C since gcc 7.  However, I found that it is the definition of `_GNU_SOURCE` that causes the APIs using `_Float##` types to be enabled.  With that it happens in C too:

```console
$ castxml -c -x c --castxml-cc-gnu-c /usr/bin/gcc-8 /usr/include/stdlib.h -D_GNU_SOURCE
/usr/include/stdlib.h:141:8: error: unknown type name '_Float32'
  141 | extern _Float32 strtof32 (const char *__restrict __nptr,
      |        ^
...
```

In C++ it happens without `-D_GNU_SOURCE` because on GNU/Linux, [g++ always defines _GNU_SOURCE](https://gcc.gnu.org/onlinedocs/libstdc++/faq.html#faq.predefined) as required by libstdc++.

**Comment by bradking** [comment]
I've released [CastXML 0.6.3](https://github.com/CastXML/CastXML/releases/tag/v0.6.3) with the fix for this.

**Comment by Nic30** [comment]
@bradking Nice!
Do you know how long does it usually take for update to propagate to deb packages of distros like Ubuntu?

**Comment by bradking** [comment]
If Ubuntu's latest gcc and latest castxml do not work together, I suggest opening an issue with them, and point at the castxml 0.6.3 release.

**Comment by badshah400** [comment]
Hi, on openSUSE, also using GCC 13.2, we see `<cmath>` complaining about:

> no known conversion from '__castxml_Float32' (aka '__castxml_Float32_s') to 'float'
> no known conversion from '__castxml_Float32' (aka '__castxml_Float32_s') to 'double'

and so on when building ITK with CastXML 0.6.3. This looks related to the issue here (if it is not, I would be happy to open a new one). Thanks for any suggestions to work around this.

```
[  825s] [1303/4670] cd /home/abuild/rpmbuild/BUILD/ITK-5.3.0/build/Wrapping/Modules/ITKCommon && /usr/bin/castxml -o /home/abuild/rpmbuild/BUILD/ITK-5.3.0/build/Wrapping/castxml_inputs/ITKCommonBase.xml --castxml-gccxml --castxml-start _wrapping_ --castxml-cc-gnu "(" /var/lib/build/ccache/bin/c++ -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -flto=auto -g -Wcast-align -Wdisabled-optimization -Wextra -Wformat=2 -Winvalid-pch -Wno-format-nonliteral -Wpointer-arith -Wshadow -Wunused -Wwrite-strings -funit-at-a-time -Wno-strict-overflow -Wno-deprecated -Wno-invalid-offsetof -Woverloaded-virtual -Wstrict-null-sentinel -fno-sized-deallocation -fno-sized-deallocation -std=c++14 ")" -w -c @/home/abuild/rpmbuild/BUILD/ITK-5.3.0/build/Wrapping/castxml_inputs/.castxml.inc /home/abuild/rpmbuild/BUILD/ITK-5.3.0/build/Wrapping/castxml_inputs/ITKCommonBase.cxx
[  825s] FAILED: Wrapping/castxml_inputs/ITKCommonBase.xml /home/abuild/rpmbuild/BUILD/ITK-5.3.0/build/Wrapping/castxml_inputs/ITKCommonBase.xml
[  825s] cd /home/abuild/rpmbuild/BUILD/ITK-5.3.0/build/Wrapping/Modules/ITKCommon && /usr/bin/castxml -o /home/abuild/rpmbuild/BUILD/ITK-5.3.0/build/Wrapping/castxml_inputs/ITKCommonBase.xml --castxml-gccxml --castxml-start _wrapping_ --castxml-cc-gnu "(" /var/lib/build/ccache/bin/c++ -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -flto=auto -g -Wcast-align -Wdisabled-optimization -Wextra -Wformat=2 -Winvalid-pch -Wno-format-nonliteral -Wpointer-arith -Wshadow -Wunused -Wwrite-strings -funit-at-a-time -Wno-strict-overflow -Wno-deprecated -Wno-invalid-offsetof -Woverloaded-virtual -Wstrict-null-sentinel -fno-sized-deallocation -fno-sized-deallocation -std=c++14 ")" -w -c @/home/abuild/rpmbuild/BUILD/ITK-5.3.0/build/Wrapping/castxml_inputs/.castxml.inc /home/abuild/rpmbuild/BUILD/ITK-5.3.0/build/Wrapping/castxml_inputs/ITKCommonBase.cxx
[  825s] In file included from /home/abuild/rpmbuild/BUILD/ITK-5.3.0/build/Wrapping/castxml_inputs/ITKCommonBase.cxx:2:
[  825s] In file included from /home/abuild/rpmbuild/BUILD/ITK-5.3.0/Modules/Core/Common/include/itkOffset.h:22:
[  825s] In file included from /home/abuild/rpmbuild/BUILD/ITK-5.3.0/Modules/Core/Common/include/itkMath.h:31:
[  825s] In file included from /usr/include/c++/13/cmath:47:
[  825s] /usr/include/math.h:1400:12: error: no matching function for call to '__iseqsigf'
[  825s]  1400 |     return __iseqsigf (__x, __y);
[  825s]       |            ^~~~~~~~~~
[  825s] /usr/include/bits/mathcalls-helper-functions.h:41:1: note: candidate function not viable: no known conversion from '__castxml_Float32' (aka '__castxml_Float32_s') to 'float' for 1st argument
[  825s]    41 | __MATHDECL_ALIAS (int, __iseqsig,, (_Mdouble_ __x, _Mdouble_ __y), iseqsig);
[  825s]       | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[  825s] /usr/include/math.h:303:3: note: expanded from macro '__MATHDECL_ALIAS'
[  825s]   303 |   __MATHDECL_1(type, function, suffix, args)
[  825s]       |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[  825s] /usr/include/math.h:299:3: note: expanded from macro '__MATHDECL_1'
[  825s]   299 |   __MATHDECL_1_IMPL(type, function, suffix, args)
[  825s]       |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[  825s] /usr/include/math.h:297:15: note: expanded from macro '__MATHDECL_1_IMPL'
[  825s]   297 |   extern type __MATH_PRECNAME(function,suffix) args __THROW
[  825s]       |               ^                                ~~~~
[  825s] /usr/include/math.h:326:34: note: expanded from macro '__MATH_PRECNAME'
[  825s]   326 | # define __MATH_PRECNAME(name,r) name##f##r
[  825s]       |                                  ^
[  825s] <scratch space>:257:1: note: expanded from here
[  825s]   257 | __iseqsigf
[  825s]       | ^
[  825s] In file included from /home/abuild/rpmbuild/BUILD/ITK-5.3.0/build/Wrapping/castxml_inputs/ITKCommonBase.cxx:2:
[  825s] In file included from /home/abuild/rpmbuild/BUILD/ITK-5.3.0/Modules/Core/Common/include/itkOffset.h:22:
[  825s] In file included from /home/abuild/rpmbuild/BUILD/ITK-5.3.0/Modules/Core/Common/include/itkMath.h:31:
[  825s] In file included from /usr/include/c++/13/cmath:47:
[  825s] /usr/include/math.h:1410:12: error: no matching function for call to '__iseqsig'
[  825s]  1410 |     return __iseqsig (__x, __y);
[  825s]       |            ^~~~~~~~~
```

**Comment by bradking** [comment]
The underlying problem is that LLVM/Clang doesn't implement GNU's builtin `_Float{32,32x,64,64x,128}` types.  With `--castxml-cc-gnu /usr/bin/g++-13` we try to preprocess the translation unit the same way `g++-13` would, so `__GNUC__` is defined to `13`, and glibc's headers assume the `_Float{32,32x,64,64x,128}` types exist, are distinct from `float`, `double`, etc., and can convert to/from the latter.  #252 taught CastXML to simulate the existence of the extra builtin distinct float types, but doesn't offer conversions.

**Comment by bradking** [comment]
I've released [CastXML 0.6.4](https://github.com/CastXML/CastXML/releases/tag/v0.6.4) with an updated fix for `<cmath>`.

**Comment by badshah400** [comment]
Many thanks, this works.
